### PR TITLE
Convert globals to classes

### LINKED ISSUE #47: Convert CMakeTest internal test representations to CMakePP objects
**Is your feature request related to a problem? Please describe.**
The internals of CMakeTest currently represent a test and its subsections as a collection of values stored in CMakePP globals. This worked just fine initially but as more and more features are added it is becoming increasingly difficult to maintain. For example, the fix for #22 was initially easy to implement, but once #24 was merged it became much more difficult to implement, due to the need to keep track of which tests and sections have been executed already as well as which sibling and parent sections still need to be executed.

**Describe the solution you'd like**
It would be much easier to add features to and maintain CMakeTest if tests and sections were instead implemented as CMakePP objects with a list of subsection objects, a reference pointing to the parent object, the name of the test or section function it represents, a boolean stating whether the section has been executed already, and any metadata like EXPECTFAIL or print_length.

### PR DESCRIPTION
**Pull request by AutonomicPerfectionist**
This PR converts the massive list of CPP globals used for storing internal test and section information into a clean CTExecutionUnit class. This does not change anything about the external API but it does clean up the internal code structure considerably and makes it far easier to add new features in the future.

Fixes #47

### CODE REVIEW

**Comment by AutonomicPerfectionist** [comment]
Since this PR changes a massive amount of internal operations, we should perform extensive integration testing before merging to ensure no regressions occur. Currently we do not have automated integration testing setup, only unit testing. Manual integration testing can be performed by switching the download URL for CMakeTest to this branch in other projects and running their unit tests to ensure the public API remained stable. CMinx and CMakePPLang would be ideal targets for this type of testing.

**Comment by AutonomicPerfectionist** [comment]
Thanks for the detailed information, I will fix the errors as soon as I can

**Comment by AutonomicPerfectionist** [comment]
@ryanmrichard @zachcran after investigating the errors I discovered several different bugs, some in CMakeTest and others in CMakePPLang. For whatever reason the bugs in Lang were not caught by previous tests, but I discovered that some `cpp_assert_signature()` calls were outdated, causing any tests that enabled debug mode to fail. I made an issue for those as well as created a pull request to fix them. I also discovered a bug in `cpp_map` and made a report for it, however I did not have the time to make the fix as it would probably require a bit more involvement, instead I renamed the offending test in my aforementioned pull request.

Additionally, there were a few naming errors in CMakeTest that cause CMakePPLang to incorrectly identify several variables as being functions instead of `desc`s, so I renamed those and pushed to this branch. After the pull request to CMakePPLang is merged all of its tests should succeed with this branch.

File: cmake_test/execution_unit.cmake [hunk 1]
```diff
@@ -0,0 +1,194 @@
+include_guard()
```
Comment (ryanmrichard) [inline cmake_test/execution_unit.cmake:1]: The formatting of this file looks like it got messed up at some point.

**Review by ryanmrichard — APPROVED** [review]

**Pull request merged**